pyriemann/sampling.py

```diff
@@ -0,0 +1,326 @@
+
+
+def sample_gaussian_spd_centered(n_samples, n_dim, sigma):
+    """Sample a Riemannian Gaussian distribution centered at the Identity
+
+    Sample SPD matrices from a Riemannian Gaussian distribution centered at the
+    Identity, which has the role of the origin in the SPD manifold, and
+    dispersion parametrized by sigma.
+
+    See https://arxiv.org/pdf/1507.01760.pdf for the mathematical details
+
+    Parameters
+    ----------
+    n_samples : int
+        how many samples to generate
+    n_dim : int
+        dimensionality of the SPD matrices to be sampled
+    sigma : float
+        dispersion of the Riemannian Gaussian distribution
+
+    Returns
+    -------
+    samples : ndarray (n_samples, n_dim, n_dim)
+        samples of the Riemannian Gaussian distribution
+    """
+
+    samples_r = sample_parameter_r(n_samples=n_samples,
+                                   n_dim=n_dim,
+                                   sigma=sigma)
+    samples_U = sample_parameter_U(n_samples=n_samples, n_dim=n_dim)
+
+    samples = np.zeros((n_samples, n_dim, n_dim))
+    for i in range(n_samples):
+        Ui = samples_U[i]
+        ri = samples_r[i]
+        samples[i] = Ui.T @ np.diag(np.exp(ri)) @ Ui
+        samples[i] = (samples[i] + samples[i].T) / 2.0  # ensure symmetry
+
+    return samples
+
+
+def sample_gaussian_spd(n_samples, Ybar, sigma):
+    """Sample a Riemannian Gaussian distribution
+
+    Sample SPD matrices from a Riemannian Gaussian distribution centered Ybar
+    and with dispersion parametrized by sigma. This distribution has been
+    defined in Said et al. "Riemannian Gaussian Distributions on the space of
+    symmetric positive definite matrices" (2016) and generalizes the notion of
+    a Gaussian distribution to the space of SPD matrices. The sampling is based
+    on a spectral factorization of SPD matrices in terms of their eigenvectors
+    (U-parameters) and the log of the eigenvalues (r-parameters).
+
+    See https://arxiv.org/pdf/1507.01760.pdf for more details
+
+    Parameters
+    ----------
+    n_samples : int
+        how many samples to generate
+    Ybar : ndarray (n_dim, n_dim)
+        center of the Riemannian Gaussian distribution
+    sigma : float
+        dispersion of the Riemannian Gaussian distribution
+
+    Returns
+    -------
+    samples : ndarray (n_samples, n_dim, n_dim)
+        samples of the Riemannian Gaussian distribution
+    """
+
+    n_dim = Ybar.shape[0]
+    samples_centered = sample_gaussian_spd_centered(
+                        n_samples,
+                        n_dim=n_dim,
+                        sigma=sigma)
+
+    # apply the parallel transport to Ybar on each of the samples
+    samples = np.zeros((n_samples, n_dim, n_dim))
+    for i in range(n_samples):
+        samples[i] = sqrtm(Ybar) @ samples_centered[i] @ sqrtm(Ybar)
+        samples[i] = (samples[i] + samples[i].T) / 2.0  # ensure symmetry
+
+    return samples
+
+
+def generate_random_spd_matrix(n_dim):
```

Comment:
It could be useful to add a parameter `n_samples` in this function: `generate_random_spd_matrix(n_samples, n_dim)`